File: BOJ/BFS&DFS/BJ2606-3.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const [N, M, ...graph] = input;
const V = Number(N);
const E = Number(M);

const parent = new Array(V + 1).fill(1).map((v, i) => i);

const find = node => {
  if(node === parent[node]) return node;
  return parent[node] = find(parent[node]);
};

const merge = (nodeA, nodeB) => {
  const pNodeA = find(nodeA);
  const pNodeB = find(nodeB);

  if(pNodeA === 1) {
    parent[pNodeB] = pNodeA;
  } else {
    parent[pNodeA] = pNodeB;
  }

  for(let i=1; i<=V; i++) find(i);
};

for(let i=0; i<E; i++) {
  const [from, to] = graph[i].split(' ').map(Number);
  
  merge(from, to);
}

console.log(parent.filter(v => v === 1).length - 1);

File: BOJ/BFS&DFS/BJ2606-3.cpp
#include <iostream>
using namespace std;

int V, E, cnt;
int parent[101];

int find(int u) {
  if(parent[u] == u) return u;
  return parent[u] = find(parent[u]);
}

void merge(int u, int v) {
  int uu = find(u);
  int vv = find(v);

  if(uu == 1) parent[vv] = uu;
  else parent[uu] = vv;

  for(int i=1; i<=V; i++) find(i);
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  cin >> V >> E;

  for(int i=1; i<=V; i++) {
    parent[i] = i;
  }

  int a, b;
  for(int i=0; i<E; i++) {
    cin >> a >> b;
    merge(a, b);
  }

  for(int i=1; i<=V; i++) {
    if(parent[i] == 1) {
      cnt++;
    }
  }
  cout << cnt-1 << '\n';
  return 0;
}